design/client-initiated-backchannel-authentication-flow.md

```diff
@@ -0,0 +1,494 @@
+* Client ID : ID of CD who sent Backchannel Authentication request
+* Scope     : OAuth2's scope which CD requested to get consent from the end-user
+
+This information can be retrieved from CIBA Flow's context date identified by "auth_req_id" which is related to received "Auth Result ID".
+
+AuthN/AuthZ by AD part does not communicate with Token Req/Res part directly because the latter is invoked by CD's token request. Therefore, keycloak implements Auth Result Cache to hold the result of AuthN/AuthZ by AD. AuthN/AuthZ by AD part pushes this result to this store with "Auth Result ID" as the access key. Token Req/Res part retrieves this result from this store with "Auth Result ID".
+
+#### From Token Req/Res part to AuthN/AuthZ by AD part
+
+CIBA protocol specification does not describe whether Token Request affects AuthN/AuthZ by AD.
+
+Therefore, no information needs to be passed from Token Req/Res part to AuthN/AuthZ by AD part.
+
+### Default Implementation of AuthN/AuthZ by AD part of Outside CIBA Protocol
+
+There are several ways of AuthN/AuthZ by AD (e.g. push notification to an android device, push notification to an iOS device, send SMS message to an android device, send SMS message to an iOS device, make a telephone call, etc.). Each of such the way uses different device and technology so that keycloak cannot cover all of such ways. 
+
+Considering this point, this default implementation delegates AuthN/AuthZ by AD to other entity called here "Decoupled Auth Server" outside keycloak. It assumes that this Decoupled Auth Server is under control by an administrator of keycloak.
+
+```
+---------------------------------+                    +-----------+
+| keycloak                       |                    | Decoupled |
+|  +--------------------------+  |                    | Auth      |
+|  | Outside CIBA             |  |                    | Server    |
+|  |  +--------------------+  |  | (i) POST           |           |
+|  |  | AuthN/AuthZ by AD  | --[Decoupled Auth ID]--> |           |
+|  |  |                    | <-[Decoupled Auth ID]--- |           |
+|  |  +--------------------+  |  |          (ii) POST |           |
+|  +------------|-------------+  |                    |           | 
+|         (iii) |                |                    |           |
+|         [Auth Result ID]       |                    |           |
+|               V                |                    |           |
+|     +--------------------+     |                    |           |
+|     | Auth Result Cache  |     |                    |           |
+|     |                    |     |                    |           |
+|     +--------------------+     |                    |           |
++--------------------------------+                    +-----------+
+```
+
+(i)   : keycloak sends to Decoupled Auth Server an AuthN/AuthZ by AD request with "Decoupled Auth ID" bound with "Auth Result ID" bound with "auth_req_id" issued to CD.
```

Comment:
It is not clear for me yet why we need the `Decoupled Auth ID`. I have an idea, but maybe you can provide more details (not expose `auth_req_id`, for instance)

Also, don't you also want to send the `binding_message` so that the AD can show that information when obtaining the authorization from the user?

Reply:
As for `Decoupled Auth ID`, I plan that it is version 4 UUID String and keycloak internally maps it to corresponding `auth_req_id` so that `auth_req_id` itself is not exposed to Decoupled Auth Server. If you have a good idea, I would be happy if you tell me.

As for `binding_message`, yes I agree with you. I will describe the interface the same as "Interface between Inside and Outside CIBA Protocol" section.

Reply:
I see. That is what I was suspecting in regards to having the decoupled auth id.

But how bad is it to communicate the `auth_req_id` itself if this data is transmitted over a secure protocol? Which I think is a MUST ...

Reply:
Yes, as you said and considering the followings point, I would like to use `auth_req_id` instead of `Decoupled Auth ID` :

- keycloak and Decoupled Auth Server are under the control by the same adnimistrator (but keycloak cannot check it. keycloak's user themselvles confirms it)

-  communcation between keycloak and Decoupled Auth Server must be over TLS

- `auth_req_id` is bound with CD so that Decoupled Auth Server cannot receive tokens in return for `auth_req_id`

WDYT?

Reply:
For me it makes sense. Even if the decoupled server is not within the same domain, he can't do much about it. And in the worst case, we can allow passing the `Decoupled Auth ID` as you initially proposed. Which would be just an opaque reference to the `auth_req_id`.